### ISSUE TITLE
Virtual Links nested in Absent/Not Link not working 

### ISSUE THREAD

**Issue by AmeBel** [opencog/opencog#1520]
The following is returning an empty SetLink,

Input
```
(define pln-rule-deduction
    (BindLink
        (VariableList
            (TypedVariableLink
                (VariableNode "$A")
                (TypeNode "ConceptNode")
            )
            (TypedVariableLink
                (VariableNode "$B")
                (TypeNode "ConceptNode")
            )
            (TypedVariableLink
                (VariableNode "$C")
                (TypeNode "ConceptNode")
            )
        )
        (ImplicationLink
            (AndLink
                (InheritanceLink
                    (VariableNode "$A")
                    (VariableNode "$B")
                )
                (InheritanceLink
                    (VariableNode "$B")
                    (VariableNode "$C")
                )
                ; To avoid matching (Inheritance A B) and (Inheritance B A)
                (AbsentLink
                    (EvaluationLink
                        (GroundedPredicateNode "scm: cog-equal?")
                        (ListLink
                            (VariableNode "$A")
                            (VariableNode "$C")
                        )
                    )
                )
            )
            (ExecutionOutputLink
                (GroundedSchemaNode "scm: pln-formula-simple-deduction")
                (ListLink
                    (InheritanceLink
                        (VariableNode "$A")
                        (VariableNode "$B"))
                    (InheritanceLink
                        (VariableNode "$B")
                        (VariableNode "$C")
                    )
                    (InheritanceLink
                        (VariableNode "$A")
                        (VariableNode "$C")
                    )
                )
            )
        )
    )
)


; -----------------------------------------------------------------------------
; Check whether two nodes are equal.
;
; If they are equal then it will return TRUE_TV else it returns FALSE_TV.
; -----------------------------------------------------------------------------
(define (cog-equal? atom-1 atom-2)
    (if (equal? atom-1 atom-2)
        (stv 1 1)
        (stv 0 1)
    )
)

```

Command run
```
guile> (load-scm-from-file "../opencog/python/pln/examples/deduction/atomspace_contents.scm")
guile> (cog-bind pln-rule-deduction)
(SetLink
)
```

What am I missing ?

**Comment by linas** [comment]
Yes, this is a real bug. The virtual-link code currently ignores the absent-link mechanism, mostly because making the two work-together was hard,  I can take a look at this tomorrow

**Comment by AmeBel** [comment]
thanks

**Comment by AmeBel** [comment]
Just tested #1522 , not working

**Comment by misgeatgit** [comment]
what do you mean by 'To avoid matching (Inheritance A B) and (Inheritance B A)' ? I mean I don't see any logical error in the output without the AbsentLink being defined in the example you provided.

**Comment by williampma** [comment]
Including (Inheritance A B) and (Inheritance B A) will end up creating (Inheritance A A) where $C is matched to the same thing as $A

**Comment by misgeatgit** [comment]
I added that kind of atom in the the example provided and 
               (AbsentLink                      
                        (InheritanceLink
                            (VariableNode "$B")
                            (VariableNode "$A")                       
                    )
seems to solve the problem.

**Comment by williampma** [comment]
But what if we have these in the atomspace
```
InhLink dog cat
InhLink cat dog
InhLink cat mouse
```
where A = dog, B = cat, C = mouse.  We might still want to output `InhLink dog mouse` even though `InhLink cat dog` is not absent.

**Comment by misgeatgit** [comment]
OK then, pln-formula-simple-deduction argument can be modified to take only the clauses and creates Inh $A $C with in its defintion only when $A and $B are not equal.

**Comment by AmeBel** [comment]
@misgeatgit that wouldn't prevent inh $A $A from being formed.

**Comment by AmeBel** [comment]
@misgeatgit  oh i see what you mean, now. In any case this bug/lack-of-feature was discovered when working with pln-deduction, so your suggestions could be used, with another headache passed to the backward chainer. In any case, the feature would be handy to have.

**Comment by linas** [comment]
OK, so I just pushed a fix for this in push request #1550.  Please test.  Not however, the following important changes:

-- you should be using NotLink, instead of AbsentLink.  There is a difference: AbsentLink says: "this sub-pattern should not be present in the atomspace"   while NotLink says "this GPN should be false"  so for this example, you really want NotLink.

I recently added an EqualLink, which is short-hand for a GPN that does an atom-equality test.  It should give same results as scm: cog-equal? but should be a bit faster, since its not a scheme call.

I added EqualLink because it converts the black-box of "cog-equal?" into something "obvious", that one can reason about directly. I (we) can and should add more of these kinds of predicates, so as to make reasoning easier and also make a MOSES port easier.  Common 'black boxes" really should be well-defined link types.  I have some ideas for more; we need a separate discussion for this, in some other feeature request, though.

**Comment by AmeBel** [comment]
Tested using GPN and EqualLink works :+1:  Thanks @linas
